### DISCUSSION TITLE
Pickled data in ows file

### DISCUSSION THREAD

**Discussion by mw25** [biolab/orange3#5882] [Q&A] [answered]
> You can open the file (it's just an XML) and see for yourself what's in there. Every widget saves its internal data;  some widgets store it as literals (**we plan to migrate all widgets to that format eventually**), while for others the data is base64 encoded pickle. You can `base64decode` such text and give it to `pickle.loads`. But as you'll see, it essentially includes widget's internal variables, so the format can change at any time.

_Originally posted by @janezd in https://github.com/biolab/orange3/discussions/5682#discussioncomment-1588324 - Highlighting inserted by me._

I am experimenting with Orange.canvas.run or the 'batch runner' and would like to change the widget settings (at least some like the file path of the file widget) directly in the ows file.
This is a bit more difficult with pickled settings....

So as a feedback for you: 
I would be very happy if all settings in the ows file are saved as literals.

Do you have a rough idea if at all and when this could be implemented?

**Comment by janezd** [comment, marked as answer]
Widgets can store any data, much of which cannot be directly expressed as literals. https://github.com/biolab/orange-widget-base/pull/62 introduced settings annotations, which allowed any data (within some reasonable limitations) to be converted to literals and back when saving and loading. The change was substantial and cut through all widgets. Making it backward compatible -- because of add-ons and other projects based on Orange, and to be able to read existing workflows -- added another nasty challenge.

https://github.com/biolab/orange-widget-base/pull/62 was reverted ten days after it was merged, because a number of issues and bugs popped up.

This was exactly a year ago. It will have to be completed some day, but I'm scared to touch it again, certainly not during semester. And even in summer I prefer cycling in forests full of bears, rather than navigating through settings. Settings are scarier.

We also had some discussions about simplifying the settings mechanism. This could somewhat simplify certain aspects.

In short: we are aware of the problem, but I can't give you any schedule.

In shorter: just unpickle from XML and change the settings. :)
